File: built-in plugins dsp/mb-restore-depop.cpp
/**
 * MB Pop Remover
 * Category : effect
 * Type     : gate
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Detect and attenuate plosive pops from vocal recordings
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_RESTORE_DEPOP_H
#define MB_RESTORE_DEPOP_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbRestoreDepop : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-restore-depop";
    static constexpr const char* PLUGIN_NAME    = "MB Pop Remover";
    static constexpr const char* PLUGIN_TYPE    = "gate";
    static constexpr const char* PLUGIN_CATEGORY = "effect";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float sensitivity = 0.5f;  // range [0, 1]
    float frequency = 100f;  // range [30, 300]
    float reduction = 12f;  // range [0, 24]
    float speed = 0.5f;  // range [0, 1]
    };

    MbRestoreDepop() = default;
    ~MbRestoreDepop() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.sensitivity = std::clamp(params.sensitivity, 0f, 1f);
        params.frequency = std::clamp(params.frequency, 30f, 300f);
        params.reduction = std::clamp(params.reduction, 0f, 24f);
        params.speed = std::clamp(params.speed, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Pop Remover
        return input;
    }
};

#endif // MB_RESTORE_DEPOP_H

File: built-in plugins dsp/mb-restore-depop.ts
import type { PluginDefinition } from "../server/services/pluginHostService";

const MbRestoreDepopPlugin: PluginDefinition = {
  id: "mb-restore-depop",
  slug: "mb-restore-depop",
  name: "MB Pop Remover",
  category: "effect",
  type: "gate" as any,
  version: "1.0.0",
  description: "Detect and attenuate plosive pops from vocal recordings",
  author: "Max Booster",
  grade: "A",
  parameters: [
    {
      id: "sensitivity",
      name: "Sensitivity",
      type: "float",
      defaultValue: 0.5,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
    {
      id: "frequency",
      name: "Focus Frequency",
      type: "float",
      defaultValue: 100,
      minValue: 30,
      maxValue: 300,
      automatable: true,
    },
    {
      id: "reduction",
      name: "Reduction",
      type: "float",
      defaultValue: 12,
      minValue: 0,
      maxValue: 24,
      automatable: true,
    },
    {
      id: "speed",
      name: "Recovery Speed",
      type: "float",
      defaultValue: 0.5,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
  ],
  defaultPreset: {
    sensitivity: 0.5,
    frequency: 100,
    reduction: 12,
    speed: 0.5,
  },
};

export default MbRestoreDepopPlugin;
